### PR TITLE
[13.0][ADD] new module: server_ftp

### PR DESCRIPTION
**Pull request by ntsirintanis**
This module enables connections via FTP, SFTP, FtpLs (wip).
Operations such as retrieving a file, checking if a file exists, are all centralized in an Abstract python class, so the user can operate without bothering with what protocol they are using. For more exquisite operations, the user may retrieve a pure connection object in a straighforward way.

### CODE REVIEW

**Comment by ntsirintanis** [comment]
@NL66278 @hbrunn @gfcapalbo @lfreeke

File: server_ftp/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Copyright 2020 Therp BV <https://therp.nl>
```
Comment (NL66278) [inline server_ftp/__init__.py]: No copyright in init files!

File: server_ftp/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2020 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Server FTP",
+    "version": "13.0.1.0.0",
+    "author": "Therp BV, Odoo Community Association (OCA)",
+    "website": "http://github.com/OCA/server-tools",
```
Comment (NL66278) [inline server_ftp/__manifest__.py]: https:

File: server_ftp/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    "depends": ["base"],
+    "data": [
+        "security/ir.model.access.csv",
+        "views/server_ftp_views.xml",
+        "data/server_ftp_data.xml",
+        "data/server_ftp_folder_data.xml",
+    ],
+    "external_dependencies": {"python": ["pysftp"]},
+    "installable": True,
+    "application": True,
```
Comment (NL66278) [inline server_ftp/__manifest__.py]: It is not an application but an enabling module.

File: server_ftp/views/server_ftp_views.xml [hunk 4]
```diff
@@ -0,0 +1,127 @@
+<?xml version="1.0" ?>
```
Comment (NL66278) [inline server_ftp/views/server_ftp_views.xml:1]: Leave out _views from the name of this file. It is already in a views folder.

File: server_ftp/data/server_ftp_data.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
Comment (NL66278) [inline server_ftp/data/server_ftp_data.xml]: Make all this into demo. Actually not all of this. One ftp server with two folders is quite enough. Put in a realisticly looking url like ftp://ftp.example.com

File: server_ftp/lib/ftp_server.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2020 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from ftplib import FTP
+
+from .abstract_ftp_server import AbstractFtpServer
+
+
+class FtpServer(AbstractFtpServer):
```
Comment (NL66278) [inline server_ftp/lib/ftp_server.py]: In class names and other places: Use the fully capitalized FTP. Like FTPServer. This is closer to all the underlying code, and therefore easier to remember and maintain.

File: server_ftp/data/server_ftp_data.xml (hunk 5 above)
Comment (ntsirintanis) [reply server_ftp/data/server_ftp_data.xml]: Actually we need to change url into host, as the argument is a hostname and not an url

File: server_ftp/lib/abstract_ftp_server.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        """ Connect to object """
+        raise NotImplementedError
+
+    @abstractmethod
+    def close(self):
+        """ Close connection """
+        raise NotImplementedError
+
+    @abstractmethod
+    def put(self, file_like, name):
```
Comment (NL66278) [inline server_ftp/lib/abstract_ftp_server.py]: As put is not used in FTP but is used for storing local files mentioned by name in pysftp, this is confusing for people knowing pysftp. Stick to putfo() for this one as well. By the way, do not forget the getfo() and its implementations.

File: server_ftp/lib/abstract_ftp_server.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from abc import ABC, abstractmethod
+
+
+class AbstractFtpServer(ABC):
+    def __init__(self, url, user, password, port, path):
```
Comment (NL66278) [inline server_ftp/lib/abstract_ftp_server.py]: Do not connect in the init method. just put self.server = None.

So connection is always done by calling the connect method.

File: server_ftp/lib/abstract_ftp_server.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        self.server = self.connect(url, user, password, port, path)
+        super().__init__()
+
+    def get_server(self):
+        return self.server
+
+    @abstractmethod
+    def connect(self, url, user, password, port, path):
```
Comment (NL66278) [inline server_ftp/lib/abstract_ftp_server.py]: Do not change path in the abstract connect method or its implementations.

File: server_ftp/lib/abstract_ftp_server.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
+# Copyright 2020 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from abc import ABC, abstractmethod
+
+
+class AbstractFtpServer(ABC):
```
Comment (NL66278) [inline server_ftp/lib/abstract_ftp_server.py]: AbstractFTPServer

File: server_ftp/lib/ftp_server.py (hunk 6 above)
Comment (NL66278) [inline server_ftp/lib/ftp_server.py]: FTPServer

File: server_ftp/lib/sftp_server.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
+# Copyright 2020 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+import pysftp
+
+from .abstract_ftp_server import AbstractFtpServer
+
+
+class SftpServer(AbstractFtpServer):
```
Comment (NL66278) [inline server_ftp/lib/sftp_server.py]: SFTPServer

File: server_ftp/models/__init__.py [hunk 12]
```diff
@@ -0,0 +1,4 @@
+# Copyright 2020 Therp BV <https://therp.nl>
```
Comment (NL66278) [inline server_ftp/models/__init__.py]: The usual nag about not having copyright lines in init files.

File: server_ftp/models/server_ftp.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
+# Copyright 2020 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import _, exceptions, fields, models
+
+from ..lib.ftp_server import FtpServer
+from ..lib.sftp_server import SftpServer
+
+
+class ServerFtp(models.Model):
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py]: ServerFTP

File: server_ftp/models/server_ftp.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    _name = "server.ftp"
+    _description = "Server ftp"
+
+    name = fields.Char(required=True)
+    url = fields.Char(string="URL")
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py]: Should also be required

File: server_ftp/models/server_ftp.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    user = fields.Char(string="Username")
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py]: Make it required with default='anonymous'.

File: server_ftp/models/server_ftp.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    password = fields.Char()
+    port = fields.Integer(default=21)
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py]: Also required

File: server_ftp/models/server_ftp.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    server_type = fields.Selection(
+        selection=[("ftp", "FTP"), ("ftpls", "FtpLs"), ("sftp", "SFTP")],
+        string="Connection protocol",
+    )
+    state = fields.Selection(
+        selection=[("draft", "Not Confirmed"), ("done", "Confirmed")],
+        default="draft",
+        readonly=True,
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py:34]: Also required, so it will have NOT NULL in the database.

File: server_ftp/models/server_ftp.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    _description = "Server ftp"
+
+    name = fields.Char(required=True)
+    url = fields.Char(string="URL")
+    user = fields.Char(string="Username")
+    password = fields.Char()
+    port = fields.Integer(default=21)
+    server_type = fields.Selection(
+        selection=[("ftp", "FTP"), ("ftpls", "FtpLs"), ("sftp", "SFTP")],
+        string="Connection protocol",
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py:28]: Also required. Make FtpTls into FTP_TLS to reflect the actual class name.

File: server_ftp/readme/CONFIGURE.rst [hunk 19]
```diff
@@ -0,0 +1,7 @@
+To configure this module, you need to:
+
+#. Go ``to Settings/Technical/FTP/Servers``
+#. Find your favourite protocol and configure connection settings
+#. In the form view of the record, edit file structure.
+#. If more folders are needed, just add data records with a proper xml_id
```
Comment (NL66278) [inline server_ftp/readme/CONFIGURE.rst:6]: This is for telling users how to configure, so leave out the lines about data records.

File: server_ftp/readme/DESCRIPTION.rst [hunk 20]
```diff
@@ -0,0 +1,9 @@
+This module enables connections via FTP, SFTP, FtpLs (wip).
+Operations such as retrieving a file, checking if a file exists, are all centralized
+in an Abstract python class, so the user can operate without even noticing what protocol
+they are using. For more exquisite operations, the user may retrieve a pure connection object in a straighforward way.
+
+TODOS:
+* Implement FtpLS
```
Comment (NL66278) [inline server_ftp/readme/DESCRIPTION.rst]: This should be trivial. Just base the FTPTLSServer class on FTPServer class, but override the connect method.

File: server_ftp/lib/ftp_server.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+        server = FTP()
+        # Init step by step
+        server.connect(host=url, port=port)
+        server.login(user=user, passwd=password)
+        server.cwd(path)
+        return server
+
+    def close(self):
+        """ Close connection """
+        self.server.quit()
```
Comment (NL66278) [inline server_ftp/lib/ftp_server.py:19]: And self.server = None

**Review by NL66278 — CHANGES_REQUESTED** [review]
Nice addition. Some changes would be helpfull. See other comments.

File: server_ftp/models/server_ftp_folder.py [hunk 22]
```diff
@@ -0,0 +1,16 @@
...
+    _name = "server.ftp.folder"
+    _description = "Server ftp folder"
+
+    name = fields.Char()
+    path = fields.Char(help="change path upon connection")
+    server_id = fields.Many2one(comodel_name="server.ftp")
+
+    def connect(self):
+        """ Return appropriate FTP class, cd into path"""
+        return self.server_id.connect(path=self.path)
```
Comment (NL66278) [inline server_ftp/models/server_ftp_folder.py]: First connect, then cd, then return server

File: server_ftp/models/server_ftp.py [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+            raise exceptions.Warning(_("Make sure that required info is set"))
+        try:
+            server = self.connect()
+        except Exception as e:
+            raise exceptions.Warning(str(e))
+        else:
+            server.close()
+            self.state = "done"
+
+    def connect(self, path="/"):
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py]: Leave out path here. (And in abstract class and implementations).

File: server_ftp/readme/DESCRIPTION.rst (hunk 20 above)
Comment (ntsirintanis) [reply server_ftp/readme/DESCRIPTION.rst]: Yeah, basically it is already there (well not here there, but somewhere in my forked repo). Just haven't tested it yet

**Comment by hbrunn** [comment]
I think you can save a lot of boilerplate by migrating https://github.com/OCA/server-backend/tree/12.0/base_external_system and build the specific file transfer things on top of it. This then also integrates better with other OCA stuff.

What you absolutely should change is all names: All this code talks to a server, but is in the role of a client itself. As well the module name as some of your variable/class names should reflect that.

Further, the readme doesn't explain what the folder configurations are good for. This might be some specific use case you better implement in a private module, not in the general help-with-data-transfer-stuff-module.

**Comment by florian-dacosta** [comment]
Seems a lot like it should be integrated with https://github.com/oca/storage no?

**Comment by NL66278** [comment]
@hbrunn The folder names are similar to the folders in fetchmail_attach_from_folder. There are often specific folders where you want to put stuff, depending on the kind of stuff, and other folders where you want to get stuff, depending on what you want. You (almost) never want to connect to an ftp server just for the fun of it, but because you want to get or store files. But you might have several locations on the server for different purposes.

**Comment by NL66278** [comment]
Also whether you use Server or Client is IMHO a question of taste or style. The Classes are for a client, but they represent a server. So what do you want to stress? But if we take indeed the bolierplate from base_external_system to connect we might have an FTPSystem, an SFTPSystem etc. avoiding this choice.

**Comment by hbrunn** [comment]
@florian-dacosta interesting, never really looked into this repo. So my understanding is you say that most of the above is already done by `storage_backend_sftp`, and `storage_backend_ftp` should be added?

**Comment by florian-dacosta** [comment]
@hbrunn I did not take a deep look in this PR, but from the description, I guess all is almost done in storage repo yes.
`storage_bakend` is a base module, then we have 1 module per protocol, and indeed, ftp is missing.
Se we would just need to make `storage_backend_ftp`, which would be kind of a copy of `storage_backend_sftp` (quite small module).

These `storage_backend_**` module really just care about the file exchanging. In case of importing file into odoo, it does not manage the file storage for instance. I am not sure it is needed here.

But, I recently adapted a existing module doing just that, to make it depend on storage (instead of re-implementing ftp, sftp exchange...) : 
https://github.com/OCA/server-tools/pull/1757
(Work is done, but test are still missing and maybe some UI improvement, we should continue the work in the next weeks...)
Basically, it allows to store a external file as an attachment in odoo (and then process the file with whatever you want to do with it)
So with this last module you can store a file from any external source, as long as there is a module managing the protocol in storage repo.

Not sure it is relevant here, but I prefer to inform in case it is of interest!
I hope it is clear!

**Comment by NL66278** [comment]
I also looked on @florian-dacosta suggestion to have a look at storage_backend. Certainly an interesting set of modules. What puts me off a little bit is the number of layers and dependencies this introduces. You also will need the connector repository, the server _environment modules etc. And it does not have a centralized method to configure connection information to external systems like base_external_system. So each storage backend has its own place to put host, user, password or other connection information.

File: server_ftp/models/server_ftp.py [hunk 24]
```diff
@@ -0,0 +1,63 @@
+# Copyright 2020 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+from odoo import _, exceptions, fields, models
+
+from ..lib.ftp_server import FtpServer
```
Comment (NL66278) [inline server_ftp/models/server_ftp.py]: You have to update these import to reflect the uppercased names.

This is causing the runbot failure:
```
  File "/home/odoo/build/OCA/server-tools/server_ftp/models/server_ftp.py", line 5, in <module>
    from ..lib.ftp_server import FtpServer
ImportError: cannot import name 'FtpServer'
```

**Comment by NL66278** [comment]
This PR might be deprecated by this https://github.com/OCA/server-backend/pull/89, basing it on base_external_system, as discussed. Also using the "client" designation for he connection to the remote system.

**Comment by NL66278** [comment]
@ntsirintanis I forwarded the work done by us here: https://github.com/OCA/server-backend/pull/89, basing the ftp module on the base_external_system module as suggested, which automatically implies we will use connections as clients, answering the request on terminology. So this PR might be closed. I decided to put the base_external_system_ftp module, as I called it now, in the server-backend repo, to prevent cross dependencies between server-tools and server-backend.

File: server_ftp/models/server_ftp.py [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+        self.ensure_one()
+        server = self._make_server_instance()
+        server.connect(self.host, self.port, self.user, self.password)
+        return server
+
+    def _make_server_instance(self):
+        """Return server class for selected type."""
+        if self.server_type == "ftp":
+            return FTPServer()
+        elif self.server_type == "ftptls":
```
Comment (CasVissers-360ERP) [inline server_ftp/models/server_ftp.py:85]: Should be elif self.server_type == "ftpls":?

**Review by CasVissers-360ERP — CHANGES_REQUESTED** [review]
Nice module!

**Comment by NL66278** [comment]
@CasVissers-360ERP Thanks for your comment. But see this alternative: https://github.com/OCA/server-backend/pull/89

This loads the server-classes dynamically, so the comment on the typo is automatically taken care off...

**Comment by thomaspaulb** [comment]
Closing as the relevant work is now in another PR: https://github.com/OCA/server-backend/pull/89
